File: backend/app/schemas/activity_log.py
"""Shapes of the activity log API's responses."""
from pydantic import BaseModel

from app.schemas.common import UtcDateTime


class ActivityLogOut(BaseModel):
    activity_log_id: int
    user_id: int | None
    # Empty for actions done by the system
    user_full_name: str | None
    username: str | None
    action: str
    entity: str
    reference: str | None
    details: str | None
    created_at: UtcDateTime


class ActivityLogPage(BaseModel):
    """One page of results, plus the total, so the page can show "1–25 of 132"."""

    items: list[ActivityLogOut]
    total: int
    page: int
    page_size: int


class ActivityLogUser(BaseModel):
    user_id: int
    full_name: str
    username: str


class ActivityLogFilterOptions(BaseModel):
    """Values for the filter dropdowns: only what actually appears in the log."""

    users: list[ActivityLogUser]
    actions: list[str]
    entities: list[str]

File: frontend/src/types/activity-log.ts
// Same shapes as backend/app/schemas/activity_log.py
export type ActivityLog = {
  activity_log_id: number
  /** Empty for actions done by the system */
  user_id: number | null
  user_full_name: string | null
  username: string | null
  action: string
  entity: string
  reference: string | null
  details: string | null
  created_at: string
}

export type ActivityLogPage = {
  items: ActivityLog[]
  total: number
  page: number
  page_size: number
}

export type ActivityLogFilterOptions = {
  users: { user_id: number; full_name: string; username: string }[]
  actions: string[]
  entities: string[]
}

/** Filters for the activity log list. Leave a field out to not filter by it. */
export type ActivityLogQuery = {
  user_id?: number
  action?: string
  entity?: string
  /** Text in the reference or details */
  search?: string
  /** ISO date/time, included */
  created_from?: string
  /** ISO date/time, not included */
  created_before?: string
  page?: number
  page_size?: number
}
